omp/matrix/coo_kernels.cpp

```diff
@@ -115,48 +121,127 @@ void spmv2(std::shared_ptr<const OmpExecutor> exec,
+                    }
+                }
+                // handle row overlap with previous thread: remainder add to
+                // memory
+                for (size_type rhs = rounded_rhs; rhs < num_rhs; rhs++) {
+                    atomic_add(c->at(first, rhs),
+                               partial_sum[rhs - rounded_rhs]);
                 }
             }
+            // handle non-overlapping rows
             for (; nz < end && coo_row[nz] != last; nz++) {
                 const auto row = coo_row[nz];
                 const auto col = coo_col[nz];
-                for (size_type rhs = 0; rhs < num_rhs; rhs++) {
-                    c->at(row, rhs) += coo_val[nz] * b->at(col, rhs);
+                for (size_type rhs_base = 0; rhs_base < rounded_rhs;
+                     rhs_base += block_size) {
+#pragma unroll
+                    for (size_type i = 0; i < block_size; i++) {
+                        const auto rhs = i + rhs_base;
+                        c->at(row, rhs) +=
+                            scale * coo_val[nz] * b->at(col, rhs);
+                    }
+                }
+                for (size_type rhs = rounded_rhs; rhs < num_rhs; rhs++) {
+                    c->at(row, rhs) += scale * coo_val[nz] * b->at(col, rhs);
                 }
             }
-            for (; nz < end; nz++) {
-                const auto row = last;
-                const auto col = coo_col[nz];
-                for (size_type rhs = 0; rhs < num_rhs; rhs++) {
-                    atomic_add(c->at(row, rhs), coo_val[nz] * b->at(col, rhs));
+            if (last != sentinel_row) {
+                for (size_type rhs_base = 0; rhs_base < rounded_rhs;
+                     rhs_base += block_size) {
+                    // handle row overlap with following thread: block partial
+                    // sums
+                    partial_sum.fill(zero<ValueType>());
+                    for (auto local_nz = nz; local_nz < end; local_nz++) {
+                        const auto row = last;
+                        const auto col = coo_col[local_nz];
+#pragma unroll
+                        for (size_type i = 0; i < block_size; i++) {
+                            const auto rhs = i + rhs_base;
+                            partial_sum[i] +=
+                                scale * coo_val[local_nz] * b->at(col, rhs);
+                        }
+                    }
+                    // handle row overlap with following thread: block add to
+                    // memory
+#pragma unroll
+                    for (size_type i = 0; i < block_size; i++) {
+                        const auto rhs = i + rhs_base;
+                        const auto row = last;
+                        atomic_add(c->at(last, rhs), partial_sum[i]);
+                    }
+                }
+                // handle row overlap with following thread: block partial sums
+                partial_sum.fill(zero<ValueType>());
+                for (; nz < end; nz++) {
```

Comment:
I think the full section for the last row is the same as the first one except that you need an extra `coo_row[local_nz] == first` for the first one in two for loops. Would it make sense to create a small algorithm to treat these special cases and call it with the extra boolean condition when needed?

Reply:
That's a good observation! I'm not sure whether this is worth the effort, since it is pretty specific to the COO SpMV?